script.skinshortcuts/resources/lib/nodefunctions.py

```diff
@@ -0,0 +1,596 @@
+            message += "[CR](and %d other properties)" %( len( propertyNames ) -1 )
+        shouldRun = xbmcgui.Dialog().yesno( ADDON.getAddonInfo( "name" ), message )
+        if not shouldRun:
+            return
+
+        # Load the properties
+        currentProperties, defaultProperties = DATA._get_additionalproperties(xbmc.translatePath("special://profile/" ))
+        otherProperties, requires, templateOnly = DATA._getPropertyRequires()
+
+        # If there aren't any currentProperties, use the defaultProperties instead
+        if currentProperties == [None]:
+            currentProperties = defaultProperties
+
+        # Pull out all properties into multi-dimensional dicts
+        allProps = {}
+        allProps[ group ] = {}
+        for currentProperty in currentProperties:
+            # If the group isn't in allProps, add it
+            if currentProperty[ 0 ] not in list(allProps.keys()):
+                allProps[ currentProperty [ 0 ] ] = {}
+            # If the labelID isn't in the allProps[ group ], add it
+            if currentProperty[1] not in list(allProps[currentProperty[0]].keys()):
+                allProps[ currentProperty[ 0 ] ][ currentProperty[ 1 ] ] = {}
+            # And add the property to allProps[ group ][ labelID ]
+            if currentProperty[ 3 ] is not None:
+                allProps[ currentProperty[ 0 ] ][ currentProperty [ 1 ] ][ currentProperty[ 2 ] ] = currentProperty[ 3 ]
+
+        # Loop through the properties we've been asked to set
+        for count, propertyName in enumerate(propertyNames):
+            # Set the new value
+            log( "Setting %s to %s" %( propertyName, propertyValues[ count ] ) )
+            if len( labelIDValues ) != 1:
+                labelID = labelIDValues[ count ]
+            if labelID not in list(allProps[group].keys()):
+                allProps[ group ][ labelID ] = {}
+            allProps[ group ][ labelID ][ propertyName ] = propertyValues[ count ]
+
+            # Remove any properties whose requirements haven't been met
+            for key in otherProperties:
+                if key in list(allProps[group][labelID].keys()) and key in list(requires.keys()) and requires[key] not in list(allProps[group][labelID].keys()):
+                    # This properties requirements aren't met
+                    log( "Removing value %s" %( key ) )
+                    allProps[ group ][ labelID ].pop( key )
+
+        # Build the list of all properties to save
+        saveData = []
+        for saveGroup in allProps:
+            for saveLabelID in allProps[ saveGroup ]:
+                for saveProperty in allProps[ saveGroup ][ saveLabelID ]:
+                    saveData.append( [ saveGroup, saveLabelID, saveProperty, allProps[ saveGroup ][ saveLabelID ][ saveProperty ] ] )
+
+        # Save the new properties
+        try:
+            f = xbmcvfs.File(os.path.join(DATAPATH, xbmc.getSkinDir() + ".properties"), 'w')
```

Comment:
this supports the `with` statement in Kodi 19

Reply:
Could you explain this in more detail please. I am kind of blindly stumbling through Python development...